### ISSUE TITLE
A11y components are instrumented and in instance proxies even if a11y is not enabled

### ISSUE THREAD

**Issue by zepumph** [phetsims/joist#454]
From #445 can you think of a way to do the necessary layout in NavigationBar around the keyboardHelpButton, but only create the type if in accessibility mode.

The reason this came up is because it is in instance proxies even if the button isn't displayed.

For implementation I'm thinking something like:
```js
    // only show the keyboard help button if the sim is accessible, there is keyboard help content, and we are
    // not in mobile safari
    if ( sim.accessible && sim.keyboardHelpNode && !platform.mobileSafari ) {

      // @private - Pops open a dialog with information about keyboard navigation
      this.keyboardHelpButton = new KeyboardHelpButton( sim, sim.lookAndFeel.navigationBarFillProperty, tandem.createTandem( 'keyboardHelpButton' ) );

      this.barContents.addChild( this.keyboardHelpButton );
    }

```

Let me know what you think. Assigning to @jessegreenberg since he implemented this.

**Comment by zepumph** [comment]
Wait, all a11y things are like this. Renaming this issue to think about that issue in a larger context.

**Comment by zepumph** [comment]
Is this acceptable? Do we even have a problem here. It seems confusing that all of our phet-io sims will have ariaHerald in their instance proxies. It will be a long time before all of the phet-io sims use a11y alerts and need ariaHerald. Is this something that we just have clients ignore, and don't spend any energy on hiding these useless (when a11y no enabled) components?

**Comment by samreid** [comment]
> It will be a long time before all of the phet-io sims use a11y alerts and need ariaHerald.

Is a11y buried behind a flag, or will sims incrementally pick up a11y features (including but not limited to aria herald) from common code as we go?  If so, is this our long term plan?

**Comment by zepumph** [comment]
A11y is buried behind a flag, but the flag is omitted once a sim is ready for a11y production. I don't really know how these a11y common code components are bundled. @jessegreenberg could you please answer the question above?

**Comment by jessegreenberg** [comment]
Most A11y features are hidden behind the `accessibillity` flag, but code and implementation are not. The `accessibility` flag enables/disables creation of the accessible DOM which is what supports keyboard navigation and screen reader access. But we are generally still hitting code and instantiating things related to a11y with or without it.

At the moment I think this means that keyboard help content (sim specific), the joist/KeyboardHelpButton, joist/KeyboardHelpDialog, scenery-phet/AriaHerald, and scenery-phet/UtteranceQueue would show up as instrumented.  Should I work on preventing instantiation of these things unless a11y is enabled?

**Comment by samreid** [comment]
I'm not sure.  https://github.com/phetsims/joist/issues/445 already calls for pruning down on a lot of the a11y instrumentation, so perhaps if we are left with a small number of instrumented things that aren't used in some sims (for the near future), it is OK?  On the other hand, if we are really polishing and stabilizing a sim, it would be nice to eliminate all variables that are not usable.

**Comment by jessegreenberg** [comment]
In the case of the KeyboardHelpButton, we would need to handle layout differently. Also, there is no instance of UtteranceQueue, it is a collection of static functions where the queue is global to the sim. Hiding use of these might look like wrapping a require statement and usages with a check for the accessibility flag.

@samreid @zepumph what do you think about hiding phet-io instrumentation of a11y components behind the a11y query param?

**Comment by jessegreenberg** [comment]
Wait, query params wouldn't be enough, we would need to check `sim.accessible`.

**Comment by jessegreenberg** [comment]
> if we are left with a small number of instrumented things that aren't used in some sims (for the near future), it is OK?

To me it seems more OK for 'prototype' phase and less OK for 'stable API' phase. It sounds like we are moving toward stabilizing the PhET-iO API so I recommend removing A11y until it is ready, but would defer to opinions of the PhET-iO team. @samreid and @zepumph let me know what you think.

**Comment by samreid** [comment]
That plan sounds reasonable to me, thanks!

**Comment by zepumph** [comment]
How about adding instanceDocumentation to all of these if and only if accessibility isn't enabled, saying "This sim is not in accessible mode, this instrumented item is only for Accessibility" 

This seems like a nice solution that isn't spending time removing things just to add them back in later (since eventually all sims will have a11y. To me keeping a11y seems like the stable way to go. Thoughts?

**Comment by samreid** [comment]
It seems there are 3 levels or so.

1. Only instantiate accessibility instances when in accessibility mode
2. Always instantiate accessibility instances, but only instrument them in accessibility mode
3. Always instrument accessibility instances and instrument them, but add in different documentation when not in accessibility mode.

In each case we need logic to identify what to do when not in accessibility mode.  (1) seems most robust since one if-statement can potentially circumvent many downstream accessible instances.

**Comment by jessegreenberg** [comment]
Determining whether or not accessibility is enabled will be difficult for some instances. For example, many places where UtteranceQueue is used will not have access to the `sim` instance.

That said, `UtteranceQueue` doesn't require accessibility to be enabled for it to work, though a screen reader has to be enabled to hear it and that should probably be in  `instanceDocumentation`.  So I think it is OK that UtteranceQueue is always instrumented.

I think the more concerning cases are the KeyboardHelpButton and KeyboardHelpDialog since those cannot be accessed at all when accessibility is disabled. It should be straight forward to determine when accessibility is enabled in those cases so I think we should proceed with option 1 from https://github.com/phetsims/joist/issues/454#issuecomment-340908793 for those. @samreid and @zepumph sound OK?

**Comment by zepumph** [comment]
>That said, UtteranceQueue doesn't require accessibility to be enabled for it to work, though a screen reader has to be enabled to hear it and that should probably be in instanceDocumentation.

Shouldn't it be turned off if accessibility is turned off? If we had to republish Ohms Law right now under the phet brand with no accessibility, I  can imagine that we wouldn't want to have partially complete alerts always enabled in our production sim. That isn't a very good example, but as we move through and touch more sims with a11y, that seems important (similar to BASE phet-io release having partial keyboard nav in it last spring)

**Comment by zepumph** [comment]
In phet-io we have the ifphetio! plugin, in a11y we have `sim.accessible`. Maybe that is a good policy for all a11y code. In Utterance queue I think it could be pretty straight forward:

```js
  var accessibilityEnabled = false;
  var initialized = false;

  var UtteranceQueue = {
     initialize: function( _accessibilityEnabled ){
              initialized = true;
             accessibilityEnabled = _accessibilityEnabled;
             if( !accessibilitEnabled){ return; }
             tandem.addInstance( UtteranceQueue, TUtteranceQueue );
    }
   //....
   };

```

**Comment by jessegreenberg** [comment]
> Shouldn't it be turned off if accessibility is turned off?

I don't feel strongly, these alerts are really "invisible" unless a screen reader is enabled and without the `accessibility` flag sims are not usable with a screen reader anyway. Keyboard navigation is easier to find for most users so it is critical to hide the accessible DOM until those features are ready for production.

@zepumph in https://github.com/phetsims/joist/issues/454#issuecomment-340924853, where/when do we call `initialize`?

**Comment by zepumph** [comment]
The thought would be in the Sim constructor. I think I agree with you. Perhaps less work is more here for the alerts components. 

For the keyboard dialog, I see that it is tandemRequired. What if we made it tandemOptional. There is only 1 usage of dialog creation, in `KeyboardHelpButton` which is passed a sim instance. So we can just do:

```js
      if ( !keyboardHelpDialog ) {
       keyboardHelpDialog = new KeyboardHelpDialog( self, sim.keyboardHelpNode, {
          tandem: sim.accessible && tandem.createTandem( 'keyboardHelpDialog' )
        } );
      }

```

This just toggles the instrumentation.

**Comment by jessegreenberg** [comment]
> The thought would be in the Sim constructor.

Got it! Sorry, I thought the recommendation was for a plugin like ifphetio! and I was hesitant to do that for a11y. That does seem straight forward.  Though I think they are harmless I can see how leaking a11y features when a11y is not enabled is confusing.

> What if we made it tandemOptional.

That seems fine to me. But I am also happy to work on preventing instantiation for instances and using a pattern like https://github.com/phetsims/joist/issues/454#issuecomment-340924853 that you recommended for these statics.

**Comment by zepumph** [comment]
A list of a11y features that are run even when a11y is not enabled:

* UtteranceQueue --> (both phet-io and sim.accessible not implemented)
* AriaHerald --> ( no longer instrumented, but currently not turned off with sim.accessible )
* KeyboardHelpDialog --> ( will be solved with tandemOptional)

@jessegreenberg says that since UtteranceQueue is driven by AriaHerald, if we short circuit AH in Sim.js when a11y is not enabled, then there won't be leaking aria-live values out.

Let's try making UtteranceQueue and AriaHerald a singleton, instantiated in Sim.js. That way they won't be instrumented if `!sim.accessible`.

**Comment by zepumph** [comment]
We can use `StillnessMotionStrategy.js` as a singleton refference.

**Comment by samreid** [comment]
What about only instantiating things if accessibility is enabled?

```js
      if ( !keyboardHelpDialog && sim.accessible) {
       keyboardHelpDialog = new KeyboardHelpDialog( self, sim.keyboardHelpNode, {
          tandem: tandem.createTandem( 'keyboardHelpDialog' )
        } );
      }
```

Likewise, we can assign a global on startup (perhaps in initialize-globals) that occurrences like UtteranceQueue can use.  Perhaps if the sim isn't accessible, then UtteranceQueue.setEnabled should be set to false and the UtteranceQueue shouldn't be registered with tandem (even better if inaccessible sim doesn't register UtteranceQueue with tandem in the first place).

**Comment by zepumph** [comment]
@samreid the problem with the above code snippet is that the KeyboardHelpButton is used for layout in the navigation bar, so we need to create it even when accessibility isn't enabled. Perhaps there is a way around that.

>Likewise, we can assign a global on startup
That sounds like a good plan. I that @jessegreenberg is on that.

**Comment by samreid** [comment]
It seems odd that the KeyboardHelpButton would be used for layout even when it isn't shown.

**Comment by zepumph** [comment]
@jessegreenberg could speak to it more, we couldn't see an obvious way to factor it out.

**Comment by zepumph** [comment]
@jessegreenberg can you think of any way to only create a new instance of the Keyboard help button on the nav bar if accessibility was enabled?

Why do we need it for layout if it doesn't get added?

**Comment by zepumph** [comment]
>It seems odd that the KeyboardHelpButton would be used for layout even when it isn't shown.

I see that the help button is used for layout with the following comments:


* `// available width right of center`
* ` // horizontal positioning`
* `// title can occupy all space to the left of the PhET button`

I think that we can use conditionals to say that if there isn't a keyboardHelpButton, ignore it. We could also potentially get away with adding the keyboardHelpButton and the phetMenuButton to their own node because then we could use that in these calculations, and it wouldn't matter if the keyboardHelpButton existed or not.

**Comment by jessegreenberg** [comment]
@zepumph https://github.com/phetsims/joist/issues/454#issuecomment-345860779 sounds great, and I think @samreid mentioned that it would be an improvement.

**Comment by zepumph** [comment]
@jessegreenberg please review the above commits where I hide/disable the focusProperty, UtteranceQueue, and AriaHerald. Any suggestions? Everything a11y related is now hidden from instance proxies when phet.chipper.a11yEnabled is false. If there are other, non-phet-io recommendations please continue the conversation in https://github.com/phetsims/scenery/issues/710.

In the mean time I will work on KeyboardHelpButton refactoring.

**Comment by zepumph** [comment]
@jessegreenberg please add the above commits to the review. I think that this is about it for this issue. I no longer see any a11y code in instance proxies when a11y is not enabled. Use faradays-law, JT, and BASE as examples to see how things show up on default, and when adding the `?accessibility` flag.

**Comment by jessegreenberg** [comment]
Changes look good to me @zepumph, thanks. I think that https://github.com/phetsims/joist/commit/b40fa583ef1d181daff111cecd879e03adcfa4cc is the best way to handle layout if the KeyboardHelpButton may not exist.

I removed instrumentation from `JohnTravoltageKeyboardHelpContent`, but otherwise I couldn't find any a11y in instance proxies when a11y wasn't enabled. When a11y is enabled I see things like UtteranceQueue.

Closing, but we will be on the lookout for this in the future.
